\abschnitt{Design}
Class \ectx provides a {\bfseries low-level} and {\bfseries minimal API} on
which to build {\bfseries higher-level APIs} such as stackful coroutines
(N3985\cite{N3985}) and user-mode threads (such as Boost.Fiber\cite{bfiber}).


\uabschnitt{Suspend-by-call}
\ectxop preserves the CPU register set\footnote{defined by ABI's calling
convention}: the content of those registers is pushed at the end of the stack
of the current context (at the current stack-pointer). Then \op restores the
stack-pointer register stored in \cpp{*this} and pops the CPU register set
from the newly-restored stack.
Because the context state is preserved on the context's stack, a \ectx
instance need only store the stack-pointer register.


\uabschnitt{Call semantics}
When \ectxop is called, a new instance of \ectx is synthesized representing
the current state of the running context (e.g. the stack-pointer). This new
instance is passed to the resumed context. On initial entry, it is passed as
the first argument to the top-level function. On every subsequent resumption,
the new \ectx instance is returned by the suspended \op call.

On completion of a successful context switch, the \ectx instance on which\\
\op was called is invalidated. The data member from which the stack pointer was
just restored is set to \cpp{nullptr}.

At most one instance of \ectx can represent a given execution context. The
currently-running execution context is not represented by \emph{any} \ectx
instance. Only when \op is called on some \emph{other} \ectx instance is the
state of the running execution context captured in a synthesized \ectx
instance.

As mentioned in the \nameref{subpara:destructor}
section, \cpp{\~execution\_context<>()} on a suspended (not terminated)
instance destroys the stack managed by that instance. Thus, the stack must be
managed by only one \ectx instance.\footnote{An earlier design used reference
counting, but that subverts the intended role of this facility as an extremely
fast substrate for higher-level libraries.}

Because of the symmetric context switching (only one operation transfers
control), the target execution context must be explicitly specified.


\uabschnitt{\ectxvoid}
\label{subsec:ectxvoid}
With \ectxvoid no data will be transferred, only the context switch is
executed.

The function or lambda passed to the constructor of
a \ectxvoid must accept a single \ectxvoid parameter and return \ectxvoid.
\cppf{passing_void}
\cpp{ctx1()} resumes \cpp{ctx1}, that is, control enters the lambda passed to
the constructor of \cpp{ctx1}. Argument \cpp{ctx2} represents the previous
context: the context that was suspended by the call to \cpp{ctx1()}. When the
lambda returns \cpp{ctx2}, context \cpp{ctx1} will be terminated while the
context represented by \cpp{ctx2} is resumed, hence control returns
from \cpp{ctx1()}.


\uabschnitt{Passing data}
\label{subsec:ectxdata}
You may pass data between contexts by constructing a \ectx with template
arguments other than \cpp{void}. Consider \ectxargs (where \cpp{args...} here
represents any list of type template arguments other than \cpp{void}), for
purposes of discussion. The function or lambda that initializes that instance
must accept parameters \ectxargsargsargs.

It must return \cpp{std::execution\_context<args...>}.

The initial \ectx argument is synthesized by \op. All other arguments must be
passed explicitly to \op.

The first call to \op with those arguments populates the parameter list for
the newly-entered function or lambda.

That function or lambda switches context back to the original context by
calling the passed\\
\ectxop, passing appropriate arguments.

A \ectxargstup is returned by the \emph{original} context's call to \op. The
returned \ectx is a synthesized instance representing the context that just
suspended. The rest of the \cpp{args...} are as passed by that context to \op.

So, for instance:
\cppf{passing_single}
The \cpp{ctx1(i)} call at (a) enters the lambda in context \cpp{ctx1} with
argument \cpp{j=1}, as shown by the output at (b). The
expression \cpp{ctx2(j+1)} at (c) resumes the original context (represented
within the lambda by \cpp{ctx2}) and transfers back an integer of \cpp{j+1}.
On return from \cpp{ctx1(i)}, the assignment at (d) sets \cpp{i} to \cpp{j+1},
or 2.

The assignment at (d) illustrates a recommended idiom: since the call to \op
at (a) has invalidated \cpp{ctx1}, it should be replaced by the
newly-synthesized \ectx instance returned by \op.

To continue the example:
\cppf{passing_single_continued}
The call to \cpp{ctx1(i)} at (e) (the \emph{updated} \cpp{ctx1}) resumes
the \cpp{ctx1} lambda, returning from the \cpp{ctx2()} call at (c) and
executing the assignment at (f). Here, too, we replace the \ectx
instance \cpp{ctx2} invalidated by the \op call at (c) with the new instance
returned by that same \op call. Moreover, we replace \cpp{j} with the value
passed by the call at (e).

Finally the lambda returns (the updated) \cpp{ctx2} at (g), terminating its
context.

Since the updated \cpp{ctx2} represents the context suspended by the call at
(e), control returns to the assignment at (h). Once again we replace the
invalidated \cpp{ctx1} with the one returned by \op.

However, since context \cpp{ctx1} has now terminated, the updated \cpp{ctx1}
is invalid. Its \opbool returns \cpp{false}; its \cpp{operator\!()}
returns \cpp{true}.

This is important, since in that case the values of any remaining fields of
the returned \cpp{std::tuple} are indeterminate.

It may seem tricky to keep track of which \ectx instance is currently
valid, representing the state of the suspended context. Please bear in
mind that this facility is intended as a high-performance foundation for
higher-level libraries. It is not intended to be directly consumed by
applications.\\
We can extend the example to multiple arguments.
\cppf{passing_multiple}
\op accepts the parameters specified by \ectx's template parameters. It
returns a \cpp{std::tuple} of that \ectx specialization, prepended to those
types.


\uabschnitt{Toplevel functions: main() and thread functions}
\label{subsec:main}
\main as well as the \emph{entry-function} of a thread can be represented by an
execution context. That \ectx instance is synthesized when the running context
suspends, and is passed into the newly-resumed context.
\cppf{simple}
The \cpp{ctx1()} call at (a) enters the lambda in context ctx1.\\
The \ectx\ \cpp{ctx2} at (b) represents the execution context of \main.\\
Returning \cpp{ctx2} at (c) resumes the original context (switch back to
\main).


\uabschnitt{\ectx and std::thread}
Any execution context represented by a valid\\
\ectx instance is necessarily suspended.\\
It is valid to resume a \ectx instance on any thread -- \emph{except} that you
must not attempt to resume a \ectx instance representing \main, or
the \emph{entry-function} of some other \cpp{std::thread}, on any thread other
than its own.
\ectx provides a method to test for this.
If \cpp{std::execution\_context<>::any\_thread()} returns \cpp{false}, it is
only valid to resume that
\ectx instance on the thread on which it was initially launched.


\uabschnitt{Termination}
When you explicitly construct a particular \ectxargs specialization, passing
its constructor a function or lambda, that callable must accept that same\\
\ectxargs specialization as its first parameter. It must return that
same \ectxargs specialization as well.

When that toplevel callable returns a \ectx instance, the running context is
voluntarily terminated. Control switches to the context indicated by the
returned \ectx instance.

Returning an invalid \ectx instance (\opbool returns \cpp{false}) invokes
undefined behavior.

If the toplevel callable returns the same \ectx instance it was originally
passed (or rather, the most recently updated instance returned from the
previous instance's \op), control returns to the context that most recently
resumed the running callable. However, the callable may return (switch to)
any reachable valid \ectx instance with the correct type signature.


\uabschnitt{Exceptions}
\label{subsec:exceptions}
If an uncaught exception escapes from the toplevel context function,
\cpp{std::terminate} is called.


\uabschnitt{Invoke function on top of a context}
Sometimes it is useful to invoke a new function (for instance, to trigger
unwinding the stack) on top of a suspended context. For this purpose you may
pass to\\
\ectxop:

\begin{itemize}
  \item the special argument \cpp{invoke\_ontop\_arg}
  \item the function to execute
  \item any additional arguments required by the \ectx specialization.
\end{itemize}

The function passed in this case must accept as parameters the \ectx
specialization for that context plus any arguments specified by that
specialization. It must return the same set of arguments as the \ectx
specialization. \footnote{But in the case of \ectxvoid, the return
type is simply \cpp{void}.}\\

For purposes of discussion, consider two \ectxvoid
instances: \cpp{mctx} and \cpp{fctx}. Suppose that code running
on the program's main context instantiates \cpp{fctx} with function\\
\cpp{f()} and calls \cpp{fctx()}, thereby
entering \cpp{f()}. This is the point at which \cpp{mctx} is synthesized
and passed into \cpp{f()}.

Suppose further that after doing some work, \cpp{f()} calls \cpp{mctx()},
thereby switching context back to the main context. \cpp{f()} remains
suspended in the call to \cpp{mctx.operator()()}.

At this point the main context calls \cpp{fctx(std::invoke\_ontop\_arg, g);}
where \cpp{g()} is declared as\\
\cpp{void g();} \cpp{g()} is entered in the context of \cpp{f()}. It is as if
\cpp{f()}'s call to \cpp{mctx.operator()()} directly called \cpp{g()}.

Function \cpp{g()} has the same range of possibilities as any function called
on \cpp{f()}'s context. Its special invocation only matters when control
leaves it in either of two ways:

\begin{enumerate}
  \item If \cpp{g()} throws an exception, that exception unwinds all previous
  stack entries in that context (such as \cpp{f()}'s) as well, back to a
  matching \cpp{catch} clause.\footnote{As stated
  in \nameref{subsec:exceptions}, if there is no matching \cpp{catch} clause
  in that context, \cpp{std::terminate()} is called.}
  \item If \cpp{g()} returns, its return value becomes the value returned
  by \cpp{f()}'s suspended \cpp{mctx.operator()()} call. This is
  why \cpp{g()}'s return type must be the same as that of \op, rather than
  that of an ordinary toplevel context function.
\end{enumerate}

Consider the following example:

\cppf{ontop}

Control passes from (a) to (b) to (c), and so on.

The \cpp{ctx(std::invoke\_ontop\_arg, f2, data+1)} call at (l) passes control
to \cpp{f2()} on the context originally created for \cpp{f1()}.

The \cpp{return} statement at (n) causes the \op call at (i) to return,
executing the assignment at (o). The \cpp{int} returned by \cpp{f2()}
is directly returned to that assignment at (o).

So in this example, the call at (l) synthesizes a \ectx instance representing
the main context and updates \cpp{mctx} internally. \cpp{f2()} returns its
return value \emph{-1}. Finally, \cpp{f1()} returns its own \cpp{ctx}
variable, switching back to the main context.


\uabschnitt{Stack destruction}
\label{subsec:destruction}
On construction of \cpp{execution\_context} a stack is allocated. If the
toplevel context-function returns, the stack will be destroyed. If the
context-function has not yet returned and the \nameref{subpara:destructor} of
a valid \cpp{execution\_context} instance (\opbool
returns \cpp{true}) is called, the stack will be unwound and
destroyed.\footnote{An implementation is free to unwind the stack without
throwing an exception. However, if an exception is thrown, it should be
named \cpp{std::execution\_context\_unwind}. Portable consumer
code \emph{must} permit \cpp{std::execution\_context\_unwind} exceptions to
propagate, even if all other exceptions are caught with \cpp{catch (...)}.}

The stack on which \cpp{main()} is executed, as well as the stack implicitly
created by \cpp{std::thread}'s constructor, is allocated by the operating
system. Such stacks are recognized by \ectx, and are not deallocated by its
destructor.

\uabschnitt{Stack allocators}
\label{subsec:stackalloc}
are used to create stacks.\footnote{This concept, along with the \ectx
constructor accepting \cpp{std::allocator\_arg\_t}, is an optional part of the
proposal. It might be that implementations can reliably infer the optimal
stack representation.} Stack allocators might implement arbitrary stack
strategies. For instance, a stack allocator might append a guard page at the
end of the stack, or cache stacks for reuse, or create stacks that grow on
demand.\\
Because stack allocators are provided by the implementation, and are only used
as parameters of\\
\ectx's constructor, the StackAllocator concept is an implementation detail,
used only by the internal mechanisms of the \ectx implementation. Different
implementations might use different StackAllocator concepts.\\
However, when an implementation provides a stack allocator matching one of
the descriptions below, it should use the specified name.\\
Possible types of stack allocators:
\begin{itemize}
    \item \cpp{protected\_fixedsize}: The constructor accepts a \cpp{size\_t}
        parameter. This stack allocator constructs a contiguous stack of
        specified size, appending a guard page at the end to protect against
        overflow. If the guard page is accessed (read or write operation), a
        segmentation fault/access violation is generated by the operating
        system.
    \item \cpp{fixedsize}: The constructor accepts a \cpp{size\_t} parameter.
        This stack allocator constructs a contiguous stack of specified size.
        In contrast to \cpp{protected\_fixedsize}, it does not append a guard
        page. The memory is simply managed by \cpp{std::malloc()}
        and \cpp{std::free()}, avoiding kernel involvement.
    \item \cpp{segmented}: The constructor accepts a \cpp{size\_t} parameter.
        This stack allocator creates a segmented stack with the specified
        initial size, which grows on demand.
\end{itemize}

\uabschnitt{API}
declaration of class \ectx
\cppf{ec}
\paragraph*{member functions}
\subparagraph*{(constructor)}
constructs new execution context\\

\begin{tabular}{ l l }
    \midrule

    \cpp{execution\_context() noexcept} & (1)\\

    \midrule

    \cpp{template<typename Fn, typename ...Params>}\\
    \cpp{explicit execution\_context(Fn&& fn, Params&& ...params)} & (2)\\

    \midrule

    \cpp{template<typename StackAlloc, typename Fn, typename ...Params>}\\
    \cpp{execution\_context(std::allocator\_arg\_t, StackAlloc salloc,}\\
    \cpp{                   Fn&& fn, Params&& ...params)} & (3)\\

    \midrule

    \cpp{execution\_context(execution\_context&& other)} & (4)\\

    \midrule

    \cpp{execution\_context(const execution\_context& other)=delete} & (5)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] This constructor instantiates an invalid \ectx. Its \opbool
              returns \cpp{false}; its \cpp{operator\!()} returns \cpp{true}.
    \item[2)] takes a callable (function, lambda, object with \op) as
              argument. The callable must have signature as described
              in \nameref{subsec:ectxvoid} or \nameref{subsec:ectxdata}. The
              stack is constructed using either \cpp{fixedsize}
              or \cpp{segmented} (see \nameref{subsec:stackalloc}). An
              implementation may infer which of these best suits the code
              in \cpp{fn}. If it cannot infer, \cpp{fixedsize} will be used.
    \item[3)] takes a callable as argument, requirements as for (2). The stack
              is constructed using \emph{salloc}
              (see \nameref{subsec:stackalloc}).\footnote{This constructor,
              along with the \nameref{subsec:stackalloc} section, is an
              optional part of the proposal. It might be that implementations
              can reliably infer the optimal stack representation.}
    \item[4)] moves underlying state to new \ectx
    \item[5)] copy constructor deleted
\end{description}

{\bfseries Notes}
\newline
When a \ectx is constructed using either of the constructors accepting
\cpp{fn}, control is \emph{not} immediately passed to \cpp{fn}. Resuming
(entering) \cpp{fn} is performed by calling \cpp{operator()()} on the new
\ectx instance.\\

\subparagraph*{(destructor)}
\label{subpara:destructor}
destroys an execution context\\

\begin{tabular}{ l l }
    \midrule

    \cpp{\~execution\_context()} & (1)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] destroys a \ectx instance. If this instance represents a
              context of execution (\opbool returns \cpp{true}),
              then the context of execution is destroyed too. Specifically,
              the stack is unwound. As noted in \nameref{subsec:destruction},
              an implementation is free to unwind the stack either by
              throwing \cpp{std::execution\_context\_unwind} or by intrinsics
              not requiring \cpp{throw}.\\
\end{description}

\subparagraph*{operator=}
moves the context object\\

\begin{tabular}{ l l }
    \midrule

    \cpp{execution\_context& operator=(execution\_context&& other)} & (1)\\

    \midrule

    \cpp{execution\_context& operator=(const execution\_context& other)=delete} & (2)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] assigns the state of \cpp{other} to \cpp{*this} using move semantics
    \item[2)] copy assignment operator deleted
\end{description}

{\bfseries Parameters}
\begin{description}
    \item[other]   another execution context to assign to this object\\
\end{description}

{\bfseries Return value}
\begin{description}
    \item[*this]
\end{description}

\subparagraph*{operator()}
resume context of execution\\

\begin{tabular}{ l l }
    \midrule

    \cpp{std::tuple<execution\_context, Args ...> operator()(Args ...args)} & (1)\\

    \midrule

    \cpp{execution\_context<void> operator()()} & (2) \\

    \midrule

    \cpp{template<typename Fn>}\\
    \cpp{std::tuple<execution\_context, Args ...>}\\
    \cpp{operator()(invoke\_ontop\_arg\_t, Fn&& fn, Args ...args)} & (3)\\

    \midrule

    \cpp{template<typename Fn>}\\
    \cpp{execution\_context<void>}\\
    \cpp{operator()(invoke\_ontop\_arg\_t, Fn&& fn)} & (4)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] suspends the active context, resumes the execution context
    \item[2)] specialization of (1) for \cpp{execution\_context<void>}
    \item[3)] suspends the active context, resumes the execution context but
        executes \cpp{fn(args ...)} in the resumed context (on top of the
        last stack frame)
    \item[4)] specialization of (3) for \cpp{execution\_context<void>}
\end{description}

{\bfseries Parameters}
\begin{description}
    \item[...args] If the toplevel context-function represented
                   by \cpp{*this} has not yet been entered, the arguments
                   you pass are passed to the context-function as its
                   parameters, following the \ectx first parameter. \\
                   If the context represented by \cpp{*this} suspended by
                   calling \op, the arguments you pass
                   are constructed into a \ectxargstup and returned by
                   that suspended \op call. \\
                   See section \nameref{subsec:ectxdata}.\\
    \item[fn]      The \cpp{fn} passed to (3) must accept \cpp{Args...}. It
                   must return \cpp{std::tuple<Args...>} or simply \cpp{Arg} in
                   the case of a single argument.\\
                   The \cpp{fn} passed to (4) must accept and return \cpp{void}.\\
\end{description}

{\bfseries Return value}
\begin{description}
    \item[void]     When called on a \ectxvoid instance, \op returns
                    a different \ectxvoid instance. This new instance
                    represents the context that suspended in order to resume
                    the current context. That may or may not be the same
                    context that was previously represented by \cpp{*this},
                    depending on other context switches executed in the
                    meantime.
    \item[tuple]    When called on a \ectxargs instance, \op returns a\\
                    \ectxargstup containing a different\\
                    \ectxargs instance. This new instance represents the
                    context that suspended in order to resume the current
                    context, as above.\\
                    If the context represented by the new \ectxargs instance
                    suspended by calling \op, the arguments passed to \op are
                    used to populate the rest of the members of the
                    returned \cpp{std::tuple}.\\
                    See section \nameref{subsec:ectxdata}.\\
                    If the context represented by the new \ectxargs instance
                    voluntarily terminated by returning from its toplevel
                    context-function, the rest of the members of the
                    returned \cpp{std::tuple} are indeterminate.\\
\end{description}

{\bfseries Exceptions}
\begin{description}
    \item[1)] calls \cpp{std::terminate} if an exception escapes toplevel \cpp{fn}\\
\end{description}

{\bfseries Preconditions}
\begin{description}
    \item[1)] \cpp{*this} represents a context of execution (\opbool
              returns \cpp{true})
    \item[2)] \cpp{any\_thread()} returns \cpp{true}, or the running thread is
              the same thread on which \cpp{*this} ran previously.
\end{description}

{\bfseries Postcondition}
\begin{description}
    \item[1)] \cpp{*this} is invalidated (\opbool returns \cpp{false})
\end{description}

{\bfseries Notes}
\newline
The \emph{prologue} preserves the execution context of the calling context as
well as stack parts like \emph{parameter list} and \emph{return
address}.\footnote{required only by some x86 ABIs}
Those data are restored by the \emph{epilogue} if the calling context is
resumed.
\newline
A suspended \cpp{execution\_context} can be destroyed. Its resources will be
cleaned up at that time.
\newline
The returned \cpp{execution\_context} indicates whether the suspended context
has terminated (returned from toplevel context-function) via \opbool. If the
returned \cpp{execution\_context} has terminated, no data are transferred in
the returned tuple.
\newline
Because \op invalidates the instance on which it is called, \emph{no
valid \ectx instance ever represents the currently-running context.}
\newline
When calling \op, it is conventional to replace the newly-invalidated 
instance -- the instance on which \op was called -- with the new instance
returned by that \op call. This helps to avoid inadvertent calls to \op on the
old, invalidated instance.
\newline
For any \ectx specialization other than \ectxvoid,\\
when \op returns, it is important to test the returned \ectxargs instance using
\opbool or \cpp{operator\!()} before referencing any of the \cpp{args...} in the
returned \ectxargstup. If that context voluntarily terminated by returning
from the toplevel context-function, only the \ectxargs member of
the \cpp{std::tuple} will be populated. The rest of the members will have
indeterminate values.

\subparagraph*{operator bool}
test whether context is valid\\

\begin{tabular}{ l l }
    \midrule

    \cpp{explicit operator bool() const noexcept} & (1)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] returns \cpp{true} if \cpp{*this} represents a context of
              execution, \cpp{false} otherwise.
\end{description}

{\bfseries Notes}
\newline
A \ectx instance might not represent a context of execution for any of a
number of reasons.
\begin{itemize}
    \item It might have been default-constructed.
    \item It might have been assigned to another instance, or passed into a
          function.\\
          \ectx instances are move-only.
    \item It might already have been resumed (\op called). Calling \op
          invalidates the instance.
    \item The toplevel context-function might have voluntarily terminated the
          context by returning.
\end{itemize}
The essential points:
\begin{itemize}
    \item Regardless of the number of \ectx declarations, exactly one\\
          \ectx instance represents each suspended context.
    \item No \ectx instance represents the currently-running context.
\end{itemize}

\subparagraph*{operator!}
test whether context is invalid\\

\begin{tabular}{ l l }
    \midrule

    \cpp{bool operator\!() const noexcept} & (1)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] returns \cpp{false} if \cpp{*this} represents a context of
              execution, \cpp{true} otherwise.
\end{description}

{\bfseries Notes}
\newline
See {\bfseries Notes} for \opbool.

\subparagraph*{any\_thread}
test whether suspended context may be resumed on a different thread\\

\begin{tabular}{ l l }
    \midrule

    \cpp{bool any\_thread() const noexcept} & (1)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] returns \cpp{false} if \cpp{*this} must be resumed on the same
              thread on which it previously ran, \cpp{true} otherwise
\end{description}

{\bfseries Notes}
\newline
As stated in \nameref{subsec:main}, a \ectx instance can represent the initial
context on which the operating system runs \main, or the context created by
the operating system for a new \cpp{std::thread}.

It is not permitted to attempt to resume such a \ectx instance on any thread
other than its original thread. \cpp{any\_thread()} allows consumer code to
distinguish this case.

\subparagraph*{(comparisons)}
establish an arbitrary total ordering for \ectx instances\\

\begin{tabular}{ l l }
    \midrule

    \cpp{bool operator==(const execution\_context& other) const noexcept} & (1)\\

    \midrule

    \cpp{bool operator\!=(const execution\_context& other) const noexcept} & (1)\\

    \midrule

    \cpp{bool operator<(const execution\_context& other) const noexcept} & (2)\\

    \midrule

    \cpp{bool operator>(const execution\_context& other) const noexcept} & (2)\\

    \midrule

    \cpp{bool operator<=(const execution\_context& other) const noexcept} & (2)\\

    \midrule

    \cpp{bool operator>=(const execution\_context& other) const noexcept} & (2)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] Every invalid \ectx instance compares equal to every other
              invalid instance. But because the running context is never
              represented by a valid \ectx instance, and because every
              suspended context is represented by exactly one valid
              instance, \emph{no valid instance can ever compare equal to any
              other valid instance.}
    \item[2)] These comparisons establish an arbitrary total ordering of \ectx
              instances, for example to store in ordered containers. (However,
              key lookup is meaningless, since you cannot construct a search
              key that would compare equal to any entry.) There is no
              significance to the relative order of two instances.
\end{description}

\subparagraph*{swap}
swaps two \ectx instances\\

\begin{tabular}{ l l }
    \midrule

    \cpp{void swap(execution\_context& other) noexcept} & (1)\\

    \midrule
\end{tabular}

\begin{description}
    \item[1)] Exchanges the state of \cpp{*this} with \cpp{other}.
\end{description}
